packages/core-data/src/batch/create-batch.js

```diff
@@ -0,0 +1,187 @@
+ *                               inputs and must return a promise that
+ *                               resolves to an array of objects containing
+ *                               either `output` or `error`.
+ */
+export default function createBatch( processor = defaultProcessor ) {
+	let lastId = 0;
+	let queue = [];
+	const pending = new ObservableSet();
+
+	return {
+		/**
+		 * Adds an input to the batch and returns a promise that is resolved or
+		 * rejected when the input is processed by `batch.run()`.
+		 *
+		 * You may also pass a thunk which allows inputs to be added
+		 * asychronously.
+		 *
+		 * ```
+		 * // Both are allowed:
+		 * batch.add( { path: '/v1/books', ... } );
+		 * batch.add( ( add ) => add( { path: '/v1/books', ... } ) );
+		 * ```
+		 *
+		 * If a thunk is passed, `batch.run()` will pause until either:
+		 *
+		 * - The thunk calls its `add` argument, or;
+		 * - The thunk returns a promise and that promise resolves, or;
+		 * - The thunk returns a non-promise.
+		 *
+		 * @param {any|Function} inputOrThunk Input to add or thunk to execute.
+		 
+		 * @return {Promise|any} If given an input, returns a promise that
+		 *                       is resolved or rejected when the batch is
+		 *                       processed. If given a thunk, returns the return
+		 *                       value of that thunk.
+		 */
+		add( inputOrThunk ) {
+			const id = ++lastId;
+			pending.add( id );
+
+			const add = ( input ) =>
+				new Promise( ( resolve, reject ) => {
+					queue.push( {
+						input,
+						resolve,
+						reject,
+					} );
+					pending.delete( id );
+				} );
+
+			if ( isFunction( inputOrThunk ) ) {
+				return Promise.resolve( inputOrThunk( add ) ).finally( () => {
+					pending.delete( id );
+				} );
+			}
+
+			return add( inputOrThunk );
+		},
+
+		/**
+		 * Runs the batch. This calls `batchProcessor` and resolves or rejects
+		 * all promises returned by `add()`.
+		 *
+		 * @return {Promise} A promise that resolves to a boolean that is true
+		 *                   if the processor returned no errors.
+		 */
+		async run() {
+			if ( pending.size ) {
+				await new Promise( ( resolve ) => {
+					const unsubscribe = pending.subscribe( () => {
+						if ( ! pending.size ) {
+							unsubscribe();
+							resolve();
+						}
+					} );
+				} );
+			}
+
+			let results;
+
+			try {
+				results = await processor(
+					queue.map( ( { input } ) => input )
+				);
+
+				if ( results.length !== queue.length ) {
+					throw new Error(
+						'run: Array returned by processor must be same size as input array.'
+					);
+				}
+			} catch ( error ) {
+				for ( const { reject } of queue ) {
+					reject( error );
+				}
+
+				throw error;
+			}
+
+			let isSuccess = true;
+
+			for ( const [ result, { resolve, reject } ] of zip(
+				results,
+				queue
+			) ) {
+				if ( result?.error ) {
+					reject( result.error );
+					isSuccess = false;
+				} else {
+					resolve( result?.output ?? result );
```

Comment:
this seems to be a violation of the contract we provide for the output `Promise`. It seems possible that we added this so we could write tests that violate the input contract.

I noticed that the tests pass values like `1` and `2` and we test processors that return non-object. Are these intentionally valid use-cases of the function? if so, what are the reasons for allowing that? should we require that alternative processors conform to the `{ output } | { error }` structure?

Do we know of any alternative processors in use (apart from the core tests here)?

Reply:
Good catch, this is an undocumented shorthand for specifying a result.

I'm fine with updating the tests and removing the shorthand so that this is simply `resolve( result.output )`. 

> Do we know of any alternative processors in use (apart from the core tests here)?

I believe only the default processor is used by Gutenberg. I'd be surprised if third parties have used this API as it's pretty niche.